tests/test_typing.py

```diff
@@ -0,0 +1,137 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Test cases for the typing module."""
+
+from typing import Self
+
+import pytest
+
+from frequenz.core.typing import disable_init
+
+
+def test_disable_init_with_custom_error() -> None:
+    """Test that the custom error is raised when trying to instantiate the class."""
+
+    class CustomError(TypeError):
+        """Custom error to raise when trying to instantiate the class."""
+
+        def __init__(self) -> None:
+            """Initialize the custom error."""
+            super().__init__("Please use the factory method to create instances.")
+
+    with pytest.raises(
+        CustomError, match="Please use the factory method to create instances."
+    ):
+
+        @disable_init(error=CustomError())
+        class MyClass:
+            """Base class that doesn't provide a default constructor."""
+
+            def __init__(self) -> None:
+                """Raise an exception when parsed."""
+
+
+def test_disable_init_default_error() -> None:
+    """Test that the default error is raised when trying to instantiate the class."""
+
+    @disable_init
+    class MyClass:
+        """Base class that doesn't provide a default constructor."""
+
+        @classmethod
+        def new(cls) -> Self:
+            """Create a new instance of the class."""
+            return cls.__new__(cls)
+
+    with pytest.raises(
+        TypeError,
+        match="This class doesn't provide a default constructor, you must use a "
+        "factory method to create instances.",
+    ):
+        MyClass()
+
+    instance = MyClass.new()
+    assert isinstance(instance, MyClass)
+
+
+def test_disable_init_subclass() -> None:
+    """Test that the default error is raised when trying to instantiate a subclass."""
+
+    @disable_init
+    class MyClass:
+        """Base class that doesn't provide a default constructor."""
+
+        @classmethod
+        def new(cls) -> Self:
+            return cls.__new__(cls)
+
+    class Sub(MyClass):
+        """Subclass that doesn't provide a default constructor."""
+
+    with pytest.raises(
+        TypeError,
+        match=r".*Sub doesn't provide a default constructor, you must use a factory "
```

Comment:
would it work here just setting `match=r"Sub doesn't provide...` (without `.*`). I'm just trying to understand this

Reply:
Nevermind. I've just understood why it is needed :angel:

Reply:
The fully qualified symbol name is used (including the module) so the error will be something like `test.test_typing.Sub`. Then `Sub` is declared inside a function, the FQN gets more funky, this is why I'm not checking that part of the string.

The other test case doesn't need the `.*` in the beginning because the errors are raised at different points, where we have different info about the class being created. The one with just `Sub` is raised by the metclass when the type object (the class) is being created. It is complicated :P

But this actually makes me think that we are not using the error provided by the user always. I will have a look to see if this is really like this and if it can be fixed.

Reply:
yes, I fully understood a few minutes before your comments. Thanks anyway for the clarification and explanation